### ISSUE TITLE
ffmpeg on bintray.com

### ISSUE THREAD

**Issue by madebr** [bincrafters/community#120]
### Description of Problem, Request, or Question
For a project ([OpenRW](https://github.com/rwengine/openrw)), I want to use conan for CI and windows builds.
The project depends on OpenAL, bullet3, glm, **ffmpeg** and sdl2.

The bincrafters project has already a **ffmpeg** conan package at https://github.com/bincrafters/conan-ffmpeg.
Will this package be uploaded to bintray?

### Package Details (Include if Applicable)
* Package Name/Version: **ffmpeg/3.4**
* Operating System: **Windows**
* Compiler+version: **Visual Studio**

**Comment by SSE4** [comment]
good question - this one is pretty complex and has lots of dependencies to be done first.
however, I probably can prepare very simple version with no dependencies or very few of them.
@madebr question to you, do you need ffmpeg compiled with specific options, and do you need to support certain decoders/encoders in it (like x264, x265, vpx, etc)?

**Comment by madebr** [comment]
@SSE4
Since OpenRW is a reimplementation of a fairly old (2001) game, we don't need any recently codecs (yet).
We currently only use a few audio codecs:
- ADPCM/PCM WAV and 
- MP3 for audio

The final decision is yours.

**Comment by SSE4** [comment]
@madebr I've prepared basic build of ffmpeg:
`ffmpeg/3.4@bincrafters/stable`
it doesn't have libmp3lame dependency yet, but I am going to add this one next
feel free to try it and let me know how is it going

**Comment by madebr** [comment]
Thanks for the effort!

Using this `conanfile.txt`,
```
[requires]
openal/1.18.2@bincrafters/stable
#bullet3/2.87@bincrafters/stable     #SOON...
glm/0.9.8.5@g-truc/stable
ffmpeg/3.4@bincrafters/stable
sdl2/2.0.7@bincrafters/stable

[generators]
cmake
```
I get this error when running `conan install ..`:
```
$ conan install ../..
Version range '>=1.15' required by 'sdl2/2.0.7@bincrafters/stable' resolved to 'libiconv/1.15@bincrafters/stable'
PROJECT: Installing /home/maarten/programming/openrw/conanfile.txt
Requirements
    ffmpeg/3.4@bincrafters/stable from bincrafters
    glm/0.9.8.5@g-truc/stable from conan-center
    libalsa/1.1.5@conan/stable from conan-center
    libiconv/1.15@bincrafters/stable from bincrafters
    openal/1.18.2@bincrafters/stable from bincrafters
    sdl2/2.0.7@bincrafters/stable from bincrafters
Packages
    ffmpeg/3.4@bincrafters/stable:943d14ae58a05270707857c32a30901840465f66
    glm/0.9.8.5@g-truc/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
    libalsa/1.1.5@conan/stable:0ab9fcf606068d4347207cc29edd400ceccbc944
    libiconv/1.15@bincrafters/stable:d50a0d523d98c15bb147b18fa7d203887c38be8b
    openal/1.18.2@bincrafters/stable:31e7195ee34f1faeb5279ca4b59c8b9e97b77aa4
    sdl2/2.0.7@bincrafters/stable:f7f4876dcefdfbe819364ac6c91cd4b1a040300d

ffmpeg/3.4@bincrafters/stable: WARN: Can't find a 'ffmpeg/3.4@bincrafters/stable' package for the specified options and settings:
- Settings: arch=x86_64, build_type=Release, compiler=gcc, compiler.version=7, os=Linux
- Options: alsa=True, bzlib=False, fPIC=True, fdk_aac=False, freetype=False, iconv=False, jack=True, lzma=False, mp3lame=False, openh264=False, openjpeg=False, opus=False, postproc=True, pulse=True, sdl2=False, shared=False, vaapi=True, vdpau=True, vorbis=False, vpx=False, x264=False, x265=False, xcb=True, zlib=False, zmq=False

ERROR: Missing prebuilt package for 'ffmpeg/3.4@bincrafters/stable'
Try to build it from sources with "--build ffmpeg"
Or read "http://docs.conan.io/en/latest/faq/troubleshooting.html#error-missing-prebuilt-package"
```

It seems there is no conan ffmpeg package with hash `943d14ae58a05270707857c32a30901840465f66` at [bintray/bincrafters/ffmpeg](https://bintray.com/bincrafters/public-conan/ffmpeg%3Abincrafters#files/bincrafters%2Fffmpeg%2F3.4%2Fstable%2Fpackage)

**Comment by SSE4** [comment]
thanks for the report, I'll add GCC 7 to the builds today and let you know

**Comment by SSE4** [comment]
@madebr GCC 7 build is now available on bintray, please give another try
```
    Package_ID: 943d14ae58a05270707857c32a30901840465f66
        [options]
            alsa: True
            bzlib: False
            fPIC: True
            fdk_aac: False
            freetype: False
            iconv: False
            jack: True
            lzma: False
            mp3lame: False
            openh264: False
            openjpeg: False
            opus: False
            postproc: True
            pulse: True
            sdl2: False
            shared: False
            vaapi: True
            vdpau: True
            vorbis: False
            vpx: False
            x264: False
            x265: False
            xcb: True
            zlib: False
            zmq: False
        [settings]
            arch: x86_64
            build_type: Release
            compiler: gcc
            compiler.version: 7
            os: Linux
        outdated from recipe: False
```

**Comment by madebr** [comment]
Thanks again @SSE4 !

`Generator cmake created conanbuildinfo.cmake` (but missing bullet3)

#121 is blocking me now to confirm that everything is working well!

**Comment by madebr** [comment]
@SSE4 
To bypass #121, I've created a local `bullet` package.
Everything compiles, but linking fails:
```
[100%] Linking CXX executable ../bin/rwgame
cd /home/maarten/programming/openrw/build/build_conan/rwgame && /usr/bin/cmake -E cmake_link_script CMakeFiles/rwgame.dir/link.txt --verbose=y
/usr/bin/c++     -rdynamic CMakeFiles/rwgame.dir/GitSHA1.cpp.o CMakeFiles/rwgame.dir/main.cpp.o CMakeFiles/rwgame.dir/GameBase.cpp.o CMakeFiles/rwgame.dir/RWGame.cpp.o CMakeFiles/rwgame.dir/GameConfig.cpp.o CMakeFiles/rwgame.dir/GameWindow.cpp.o CMakeFiles/rwgame.dir/StateManager.cpp.o CMakeFiles/rwgame.dir/State.cpp.o CMakeFiles/rwgame.dir/MenuSystem.cpp.o CMakeFiles/rwgame.dir/GameInput.cpp.o CMakeFiles/rwgame.dir/states/LoadingState.cpp.o CMakeFiles/rwgame.dir/states/IngameState.cpp.o CMakeFiles/rwgame.dir/states/PauseState.cpp.o CMakeFiles/rwgame.dir/states/MenuState.cpp.o CMakeFiles/rwgame.dir/states/DebugState.cpp.o CMakeFiles/rwgame.dir/states/BenchmarkState.cpp.o CMakeFiles/rwgame.dir/DrawUI.cpp.o  -o ../bin/rwgame -Wl,-rpath,/home/maarten/.conan/data/bullet/2.87/madebr/testing/package/27cb7fbffba1a5268eda91c4da54b0254d48c3b1/lib:/home/maarten/.conan/data/openal/1.18.2/bincrafters/stable/package/31e7195ee34f1faeb5279ca4b59c8b9e97b77aa4/lib: ../lib/librwengine.a -lboost_program_options ../lib/librwlib.a -lboost_filesystem -lboost_system /usr/lib64/libGL.so /home/maarten/.conan/data/bullet/2.87/madebr/testing/package/27cb7fbffba1a5268eda91c4da54b0254d48c3b1/lib/libBulletDynamics.so /home/maarten/.conan/data/bullet/2.87/madebr/testing/package/27cb7fbffba1a5268eda91c4da54b0254d48c3b1/lib/libBulletCollision.so /home/maarten/.conan/data/bullet/2.87/madebr/testing/package/27cb7fbffba1a5268eda91c4da54b0254d48c3b1/lib/libLinearMath.so /home/maarten/.conan/data/bullet/2.87/madebr/testing/package/27cb7fbffba1a5268eda91c4da54b0254d48c3b1/lib/libBulletSoftBody.so /home/maarten/.conan/data/bullet/2.87/madebr/testing/package/27cb7fbffba1a5268eda91c4da54b0254d48c3b1/lib/libBullet3Common.so /home/maarten/.conan/data/bullet/2.87/madebr/testing/package/27cb7fbffba1a5268eda91c4da54b0254d48c3b1/lib/libBulletInverseDynamics.so /home/maarten/.conan/data/ffmpeg/3.4/bincrafters/stable/package/943d14ae58a05270707857c32a30901840465f66/lib/libavdevice.a /home/maarten/.conan/data/ffmpeg/3.4/bincrafters/stable/package/943d14ae58a05270707857c32a30901840465f66/lib/libavfilter.a /home/maarten/.conan/data/ffmpeg/3.4/bincrafters/stable/package/943d14ae58a05270707857c32a30901840465f66/lib/libavformat.a /home/maarten/.conan/data/ffmpeg/3.4/bincrafters/stable/package/943d14ae58a05270707857c32a30901840465f66/lib/libavcodec.a /home/maarten/.conan/data/ffmpeg/3.4/bincrafters/stable/package/943d14ae58a05270707857c32a30901840465f66/lib/libswresample.a /home/maarten/.conan/data/ffmpeg/3.4/bincrafters/stable/package/943d14ae58a05270707857c32a30901840465f66/lib/libswscale.a /home/maarten/.conan/data/ffmpeg/3.4/bincrafters/stable/package/943d14ae58a05270707857c32a30901840465f66/lib/libavutil.a /home/maarten/.conan/data/ffmpeg/3.4/bincrafters/stable/package/943d14ae58a05270707857c32a30901840465f66/lib/libpostproc.a -lva -lva-drm -lva-x11 -lvdpau -lX11 -lxcb -lxcb-shm -lxcb-shape -lxcb-xfixes /home/maarten/.conan/data/openal/1.18.2/bincrafters/stable/package/31e7195ee34f1faeb5279ca4b59c8b9e97b77aa4/lib/libopenal.so /home/maarten/.conan/data/libalsa/1.1.5/conan/stable/package/0ab9fcf606068d4347207cc29edd400ceccbc944/lib/libasound.a /home/maarten/.conan/data/sdl2/2.0.7/bincrafters/stable/package/f7f4876dcefdfbe819364ac6c91cd4b1a040300d/lib/libSDL2.a /home/maarten/.conan/data/sdl2/2.0.7/bincrafters/stable/package/f7f4876dcefdfbe819364ac6c91cd4b1a040300d/lib/libSDL2main.a -ldl -lrt -lpthread -lasound -ljack -lpulse -laudio -lesd /home/maarten/.conan/data/libiconv/1.15/bincrafters/stable/package/d50a0d523d98c15bb147b18fa7d203887c38be8b/lib/libiconv.a 
/usr/bin/ld: cannot find -laudio
/usr/bin/ld: cannot find -lesd
collect2: error: ld returned 1 exit status

```

**Comment by SSE4** [comment]
@madebr yep, default build is with some system packages enabled.
you're running CentOS, right? I am going to add support for it soon (currently only tested on Ubuntu), which will provide more user friendly handling of system libraries.

**Comment by madebr** [comment]
@SSE4 
I'm running fedora 26, rpm based as well.
Thanks for the support!

**Comment by SSE4** [comment]
great, I guess should be compatible with CentOS package convention. I'll let you know once CentOS/Fedora support is in place.

**Comment by madebr** [comment]
Just tried compiling using clang 4.0.1 and it fails at linking with this error:
```
/usr/bin/ld: /home/maarten/.conan/data/ffmpeg/3.4/bincrafters/stable/package/ef8c075ebd9e939a5b97466baf7b7e28815a467a/lib/libavutil.a(hwcontext_vdpau.o): undefined reference to symbol 'XDisplayString'
//usr/lib64/libX11.so.6: error adding symbols: DSO missing from command line
```

**Comment by SSE4** [comment]
@madebr thanks for the report, I'll this this missing DSO errors and let your know

**Comment by SSE4** [comment]
if it's blocking you, you can disable vdpau by adding:
`-o ffmpeg:vdapu=False` to your command line

**Comment by madebr** [comment]
```
conan install ../.. -pr clang4 -o ffmpeg:vdpau=False --build ffmpeg
```
gives the same linking error.

**Comment by SSE4** [comment]
from the same file (hwcontext_vdpau.o)?

**Comment by madebr** [comment]
Yes:
```
/home/maarten/.conan/data/ffmpeg/3.4/bincrafters/stable/package/ef8c075ebd9e939a5b97466baf7b7e28815a467a/lib/libavutil.a(hwcontext_vdpau.o)
```

**Comment by madebr** [comment]
Oh wait, this might be my error.
I thought a conan install would reinitialize the variables but I forgot CMake cached the variables...

**Comment by madebr** [comment]
@SSE4 
Yup, same error (different path to package):
```sh
/usr/bin/ld: /home/maarten/.conan/data/ffmpeg/3.4/bincrafters/stable/package/27c830ec08a0f053c83bf1b88e7b795e9781b905/lib/libavutil.a(hwcontext_vaapi.o): undefined reference to symbol 'XDisplayName'
//usr/lib64/libX11.so.6: error adding symbols: DSO missing from command line
clang-4.0: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by SSE4** [comment]
@madebr great, then pass also `-o ffmpeg:vaapi=False`. also, I forecast xcb may produce similar error, so also pass `-o ffmpeg:xcb=False`.

**Comment by madebr** [comment]
@ffmpeg
Using this install:
```sh
conan install ../.. -pr clang4 -o ffmpeg:vdpau=False -o ffmpeg:vaapi=False -o ffmpeg:xcb=False --build ffmpeg
```
This error happens during final linking:
```sh
/usr/bin/ld: /home/maarten/.conan/data/ffmpeg/3.4/bincrafters/stable/package/caeedcb56d3be7e893179c1329fd6796a67447a2/lib/libavutil.a(hwcontext_cuda.o): undefined reference to symbol 'dlclose@@GLIBC_2.2.5'
//usr/lib64/libdl.so.2: error adding symbols: DSO missing from command line
```

**Comment by SSE4** [comment]
this one looks really weird for me. first off, recipe doesn't have option for CUDA yet, so it's likely auto-detected by ffmpeg.
it's ffmpeg build fails itself, not your project linking to ffmpeg?
I can only suggest adding the following line:
`args.append('--disable-cuda')`
to the conanfile.py, line 225. can you try this locally?
I can commit fix, but it will take long time to rebuild, so better to verify if it works locally, if possible.

**Comment by madebr** [comment]
I think `libavutil` needs to link against `libdl`.

The build output of ffmpeg says this:
```sh
install prefix            /home/maarten/.conan/data/ffmpeg/3.4/bincrafters/stable/package/caeedcb56d3be7e893179c1329fd6796a67447a2
source path               .
C compiler                gcc
C library                 glibc
ARCH                      x86 (generic)
```
and
```sh
External libraries:
alsa			  jack			    libpulse		      xlib

External libraries providing hardware acceleration:
cuda			  cuvid			    nvenc		      v4l2_m2m

Libraries:
avcodec			  avfilter		    avutil		      swresample
avdevice		  avformat		    postproc		      swscale
```
Is my clang4 profile ok?
```sh
$ conan profile show clang4
Configuration for profile clang4:

[settings]
    os=Linux
    os_build=Linux
    arch=x86_64
    arch_build=x86_64
    compiler=clang
    compiler.version=4.0
    compiler.libcxx=libstdc++
    build_type=Release
[options]
[build_requires]
[env]
```

Adding `--disable-cuda` gives the following error when building ffmpeg:
```
ERROR: CUVID requires CUDA
```

**Comment by SSE4** [comment]
right, but libdl is already in the list of libraries (see method package_info, line 351).
can you also try adding `--disable-cuvid` in addition to `-disable-cuda`?
so it appears either cuda's pkg-config doesn't specify dl, or ffmpeg's configure script doesn't handle it correctly.

**Comment by madebr** [comment]
@SSE4 
From my `conan-bullet` submission, I've learnt that link order is important,
so I've moved `self.cpp.info.libs.extend(['dl', 'pthread'])` to the bottom of the elif branch.

The error now is 
```sh
/usr/bin/ld: /home/maarten/.conan/data/sdl2/2.0.7/bincrafters/stable/package/55c37e59a214e80667e9a57be194d2f2eb5080eb/lib/libSDL2.a(SDL_dynapi.c.o): undefined reference to symbol 'dlclose@@GLIBC_2.2.5
```
So it has moved to SDL2

**Comment by SSE4** [comment]
right, link order is important.
seems like the same should be applied to SDL2 recipe, can you try locally as well for it?

**Comment by madebr** [comment]
@SSE4 Ok, now I feel stupid and sorry for taking your time.
(My last finding about the order of linking was bogus because it was already appended after the libs of ffmpeg) 

The project, openrw, defines its own `FindFFmpeg.cmake` file, which of course misses a lot of flags/libs to link to the conan libraries.
Creating ALIASes (to keep it building without conan), like 
```cmake
if(USE_CONAN)
    add_library(ffmpeg INTERFACE)
    target_link_libraries(ffmpeg INTERFACE CONAN_PKG::ffmpeg)
    add_library(ffmpeg::ffmpeg ALIAS ffmpeg)
else()
    find_package(FFmpeg REQUIRED)
endif()
```
lets everything compile and link apart from a warning (by OpenAL):
```sh
/usr/bin/ld: warning: libatomic.so.1, needed by /home/maarten/.conan/data/openal/1.18.2/bincrafters/stable/package/bb59b0d6e4f149e80cfabbb438d4fffd711e0060/lib/libopenal.so, not found (try using -rpath or -rpath-link)
[100%] Built target rwgame
```
I have to install `libatomic`.
OpenRW runs fine now.

Question: some libraries of bincrafters are defaulted to shared=True, others to shared=False.
(ffmpeg has: shared=False, openal has shared=True)

**Comment by SSE4** [comment]
That’s okay. Package is still very premature and   requeuires lots of
tuning, thanks for alpha-testing it xD

For shared vs static - I strongly believe static should be sane default
choice. Shared can’t be, as not all platforms support/allow shared
libraries (e.g. iOS). Second reason, shared libraries usually need
additional handling, e.g. for installation and deployment.

**Comment by solvingj** [comment]
@SSE4 if you think openal shared should be false by default, lets make it so.

**Comment by SSE4** [comment]
@solvingj okay, I'll change it

**Comment by SSE4** [comment]
@madebr probably we need to provide our own FindFFMPEG.cmake, or at least look at this, to avoid such kind of problems. I'm adding card to remember about this.

**Comment by madebr** [comment]
Trying to compile and link on Visual Studio 2017 now
A warning displays.

The warning is
```cmd
LINK : warning LNK4098: defaultlib 'MSVCRT' conflicts with use of other libs; use /NODEFAULTLIB:library [C:\issues\conan_ffmpeg_1\build\issue.vcxproj]
```
Source at https://github.com/madebr/issues/tree/master/conan_ffmpeg_1
Appveyor output at output at https://ci.appveyor.com/project/madebr/issues/build/1.0.24/job/o3pbqbxclu4bucod

This warning does also show up when using [conan-bullet (static version)](https://ci.appveyor.com/project/madebr/issues/build/1.0.24/job/hbgvx3uoxriqi3tp) but I'm raising it anyway.

**Comment by SSE4** [comment]
@madebr do you still see warning? I support it might be caused by runtime mismatch, but if you specify runtime with `-s compiler.runtime`, it should be fine

**Comment by madebr** [comment]
@SSE4 : No I do not https://ci.appveyor.com/project/madebr/issues/build/1.0.71/job/e5li2haelpfna65l and 
https://ci.appveyor.com/project/madebr/openrw/build/1.0.434

**Comment by SSE4** [comment]
glad to hear that

**Comment by madebr** [comment]
@SSE4 Because I cannot patch `conan-ffmpeg` on travis, I'm getting the dlclose error there. (Adding `--disable-cuvid` and `--disable-cuda` to the configuration).
What would you suggest?
https://travis-ci.org/madebr/openrw/jobs/347373415

**Comment by SSE4** [comment]
@madebr I gonna disable cuda and cuvid by default in ffmpeg recipe until we add necessary options (this may require making CUDA toolkit installer, or processing system requirements, I am not sure, I'll delay it until CUDA is requested by someone). it will take some time to rebuild ffmpeg on CI, I'll let you know.

**Comment by SSE4** [comment]
@madebr could you try to update one more time? thanks

**Comment by SSE4** [comment]
I am closing this one, as all issues were addressed.
@madebr if you have any other problem(s) with ffmpeg recipe, please open new issue(s).
